### PR TITLE
model_api

### PR DESCRIPTION
**Pull request by DeNeutoy**
This is what I had in mind for the Model API. This was in the `Trainer` api PR that I am working on in tandem but I figured they are pretty much separate and the more we can quickly get an idea of what each of us has in mind for these key abstract classes, the better, so I thought i'd share.

### CODE REVIEW

File: allennlp/training/model.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
+from typing import Any, List, Dict, Union
+import torch
+
+
+class Model(torch.nn.Module):
+    """
+    This abstract class represents a model to be trained. Rather than relying on the
+    raw torch Module, we provide a slightly different API.
```
Comment (matt-gardner) [inline allennlp/training/model.py]: Some justification for _why_ we do this would be good here.

File: allennlp/training/model.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    """
+
+    def forward(self, inputs: Dict[str, torch.Tensor], compute_loss: bool = False):
```
Comment (matt-gardner) [inline allennlp/training/model.py]: I think it makes more sense to separate loss computation from `forward`.  You already have `compute_loss` as its own function - why also include it in `forward`?  Is it because the targets are included in `inputs`?  We can separate that out somehow in the data, and that'd probably be cleaner than trying to handle it like this.

File: allennlp/training/model.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        >>>         loss = self.compute_loss(outputs, targets, model_state)
+        >>>         return outputs, loss
+        >>>     else:
+        >>>         return outputs
+
+        Parameters
+        ----------
+        inputs: Dict[str, torch.Tensor], required.
+            A dictionary of tensors comprising everything needed to perform a
+            training update, _including_ labels. At inference time, simply
```
Comment (matt-gardner) [inline allennlp/training/model.py]: Emphasis is a single backtick in rST, not an underscore.

File: allennlp/training/model.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+        raise NotImplementedError
+
+    def compute_loss(self,
+                     model_output: Union[torch.Tensor, List[torch.Tensor]],
+                     targets: Union[torch.Tensor, List[torch.Tensor]],
+                     model_state: Dict[str, Any] = None) -> float:
+        """
+        Computes a scalar loss function to optimise. This method is designed
+        to be called and returned by `:func: forward` when `compute_loss = True`.
```
Comment (matt-gardner) [inline allennlp/training/model.py]: Formatting is ```:func:`forward` ```, and you need double back ticks for code.

File: allennlp/training/model.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+import torch
+
+
+class Model(torch.nn.Module):
+    """
+    This abstract class represents a model to be trained. Rather than relying on the
+    raw torch Module, we provide a slightly different API.
+    """
+
+    def forward(self,
```
Comment (matt-gardner) [inline allennlp/training/model.py]: Somehow github dropped my previous comment on this: I think it makes more sense to separate the loss computation from `forward` - it's pretty confusing to have this sometimes return the loss, and sometimes not.  What's the motivation for doing it this way?  Because the targets are included in `inputs`?  If that's the only reason, I think it'd be better to solve it by separating them in the data API somehow, instead of doing it this way.

Comment (DeNeutoy) [inline allennlp/training/model.py]: Motivations:
1) As we can build dynamic graphs, the loss function you compute may require some state from the model, which is most likely going to be generated in the `forward` function, so having the loss function created within the call to forward makes passing around this state very simple.
2) I really want to avoid having the Keras API of specifying losses and then computing them for the user, because it's so inflexible.

Instead, I guess the forward pass could set some state attribute, which then is used when calling `self.compute_loss` separately? This might be easier in the sense that your loss computation might not depend on much, so storing stuff here would be straightfoward. I guess the other thing we could have would be forward always returning a loss, which is just sometimes `None` to unify the return values?

Comment (matt-gardner) [inline allennlp/training/model.py]: Yeah, I like the last idea - just always return a value for loss, and if there are no targets, that loss is `None`.

File: allennlp/training/model.py [hunk 6]
```diff
@@ -13,9 +13,9 @@ class Model(torch.nn.Module):
     data generation, because model inputs are named.
     """
 
-    def forward(self,
+    def forward(self,  # pylint: disable=arguments-differ
                 inputs: Dict[str, torch.Tensor],
-                compute_loss: bool = False):  # pylint: disable=arguments-differ
+                compute_loss: bool = False) -> (Union[torch.Tensor, List[torch.Tensor]], Optional[float]):
```
Comment (matt-gardner) [inline allennlp/training/model.py]: I think you can also remove `compute_loss` - we just always compute the loss if the targets are given, and we return `None` if they are not.

Comment (DeNeutoy) [inline allennlp/training/model.py]: oh right and change the data api to return inputs and targets dictionaries separately or something? At the moment we don't have a distinction between them, which would make doing this hard without guessing at the names people give targets etc. Sounds good to me.

Comment (DeNeutoy) [inline allennlp/training/model.py]: This would be easy actually, just changing `Instance` and `Dataset`. We can leave the fields untouched.

Comment (matt-gardner) [inline allennlp/training/model.py]: Or we just assume that the `Model` knows which things on its inputs are targets and which are inputs, and don't make a distinction in the data.

Comment (DeNeutoy) [inline allennlp/training/model.py]: oh right, like it is now. Sorry I was confused by what you meant.

File: allennlp/training/model.py [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+            training update, `including` labels. At inference time, simply
+            pass an input dictionary which does not include labels and
+        compute_loss: bool, optional (default = False)
+            If true, this method should call :func:`compute_loss` and return
+            both the model outputs and a scalar loss to optimise.
+
+        Returns
+        -------
+
+        outputs: torch.Tensor or List[torch.Tensor]
```
Comment (matt-peters) [inline allennlp/training/model.py]: Why is the input specified as a dict but the output as a list for multiple outputs?
If the output was a dict, then you could also just insert a key for the loss to avoid having to return a tuple with second entry as `None` in the case where you don't compute the loss.

File: allennlp/training/model.py [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+        loss: float, optional
+            If ``compute_outputs=True``, return a scalar loss to optimise.
+
+        """
+
+        raise NotImplementedError
+
+    def compute_loss(self,
+                     model_output: Union[torch.Tensor, List[torch.Tensor]],
+                     targets: Union[torch.Tensor, List[torch.Tensor]],
```
Comment (matt-peters) [inline allennlp/training/model.py]: Same question regarding lists vs dict for multiple targets. Seems a dict would work better for all the reasons you are using one for `inputs` to `forward`.

File: allennlp/training/model.py [hunk 9]
```diff
@@ -0,0 +1,94 @@
+from typing import Any, List, Dict, Optional, Union
+import torch
+
+
+class Model(torch.nn.Module):
```
Comment (matt-peters) [inline allennlp/training/model.py]: Thanks for putting this together. Is the intention that a user will provide on of these `Model` implementations, specify their data with the data API, then the Trainer (or other allennlp components) will handle the training loop?

File: allennlp/training/model.py [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+
+        Parameters
+        ----------
+        model_output : torch.Tensor or List[torch.Tensor], required.
+            The output of the result of calling :func:`forward`. This
+            can be a list, as complex models may have many outputs.
+        targets : torch.Tensor or List[torch.Tensor], required.
+            The gold targets for computing some loss function with respect to
+            the model outputs.
+        model_state : Dict[str, Any], optional (default = None)
```
Comment (matt-peters) [inline allennlp/training/model.py]: I'm confused what state this might be. Is this the current model parameters, intermediate representations from lower layers that aren't part of the output or ??

File: allennlp/training/model.py [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+        outputs: torch.Tensor or List[torch.Tensor]
+            The outputs from the model.
+        loss: float, optional
+            If ``compute_outputs=True``, return a scalar loss to optimise.
+
+        """
+
+        raise NotImplementedError
+
+    def compute_loss(self,
```
Comment (matt-peters) [inline allennlp/training/model.py]: Do you have a need to call this function explicitly from outside `forward`? The way it's defined with the additional model state doesn't allow you to use it externally, unless you also provide a way to get the model state.  As a result, I'm wondering why it's a part of the public API.

File: allennlp/training/model.py [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+                     model_output: Union[torch.Tensor, List[torch.Tensor]],
+                     targets: Union[torch.Tensor, List[torch.Tensor]],
+                     model_state: Dict[str, Any] = None) -> float:
```
Comment (matt-peters) [inline allennlp/training/model.py]: Why restrict how the model represents it's internal state in this way?  Maybe it makes sense to use a different data structure... shouldn't this be a lower level detail of the model?

File: allennlp/training/model.py [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+class Model(torch.nn.Module):
+    """
+    This abstract class represents a model to be trained. Rather than relying on the
+    raw torch Module, we provide a slightly different API, particularly to align with
+    data processing code for textual data. As textual data pre-processing and deep NLP
+    model inputs tend to be more complicated than for other modalities, we use
+    dictionaries of tensors to pass inputs to the model, to keep on top of what's
+    going on. It also makes reading model implementations easier to connect to
+    data generation, because model inputs are named.
+    """
```
Comment (schmmd) [inline allennlp/training/model.py]: I would re-word this comment.  I would emphasize the benefits of this model over the PyTorch model earlier on.  Maybe:

> This abstract class represents a model to be trained.  Rather than relying on the raw torch model, we provide a richer API that maintains more metadata about the tensors used.  This makes reading model implementations easier to connect to data generation, because model inputs are named.

Still, this documentation leaves me wondering whether we should just be using the PyTorch API (I have not looked at the code yet).

File: allennlp/training/model.py (hunk 10 above)
Comment (DeNeutoy) [inline allennlp/training/model.py]: This could be the actual inputs to the model in a case where you are doing semi-supervised learning where for some of your data you have additional inputs, like POS tags etc, for which you want to augment/change your loss.

File: allennlp/training/model.py (hunk 9 above)
Comment (DeNeutoy) [inline allennlp/training/model.py]: Yep, exactly.

File: allennlp/training/model.py (hunk 7 above)
Comment (DeNeutoy) [inline allennlp/training/model.py]: fair point, I'll change this.

File: allennlp/training/model.py (hunk 11 above)
Comment (DeNeutoy) [inline allennlp/training/model.py]: I was thinking that you might want to evaluate on hold out data using the model separately, where the model state is actually provided directly, but you are right I think this is a very narrow use case; i'll make it private.

File: allennlp/training/model.py (hunk 12 above)
Comment (DeNeutoy) [inline allennlp/training/model.py]: This seems slightly contradictory to your other comments about the inputs and outputs - I guess because this is used only within the model so the user can do what they want with it? These types are only hints, but I'm ambivalent on this.

Comment (matt-gardner) [inline allennlp/training/model.py]: Because the method is now going to be private, you probably don't need this parameter at all?  If the model needs some internal state, it can just save it to `self` however it wants to.

Comment (DeNeutoy) [inline allennlp/training/model.py]: Sure, sounds good.

File: allennlp/training/model.py (hunk 10 above)
Comment (matt-peters) [inline allennlp/training/model.py]: In that case it would be passed in via the inputs, right?

File: allennlp/training/model.py (hunk 12 above)
Comment (matt-peters) [inline allennlp/training/model.py]: The difference is that the inputs and outputs have some correspondence to things specified in the data generators (so naming them by some key will presumably be necessary in the training loop), but the internal state doesn't have to have this correspondence.

File: allennlp/training/model.py [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+    wish to use an AllenNLP model inside a Container (such as nn.Sequential), you must
+    interleave the models with a wrapper module which unpacks the dictionary into
+    a list of tensors.
+
+    Finally, in order for your model to be trained using the :class:`~allennlp.training.Trainer`
+    api, the output dictionary of your Model must include a "loss" key, which will be
+    optimised during the training process.
+    """
+    #  TODO(Mark): Implement this.
+    def forward(self, *inputs) -> Dict[str, torch.Tensor]:  # pylint: disable=arguments-differ
```
Comment (matt-gardner) [inline allennlp/training/model.py]: Don't we want the input to also be a dictionary?

File: allennlp/training/model.py [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+    can be unpacked and passed into other layers. One caveat to this is that if you
+    wish to use an AllenNLP model inside a Container (such as nn.Sequential), you must
+    interleave the models with a wrapper module which unpacks the dictionary into
+    a list of tensors.
+
+    Finally, in order for your model to be trained using the :class:`~allennlp.training.Trainer`
+    api, the output dictionary of your Model must include a "loss" key, which will be
+    optimised during the training process.
+    """
+    #  TODO(Mark): Implement this.
```
Comment (matt-gardner) [inline allennlp/training/model.py]: What's this TODO for?

File: allennlp/training/model.py [hunk 16]
```diff
@@ -0,0 +1,64 @@
+from typing import Dict
+import torch
+
+
+class Model(torch.nn.Module):
+    """
+    This abstract class represents a model to be trained. Rather than relying completely
+    on the Pytorch Module, we provide a richer API that maintains more metadata
```
Comment (matt-gardner) [inline allennlp/training/model.py]: Because we're not actually providing more functions, just changing the signature of one of them, I might just say "we modify the `forward` API to make it more suitable for easy training." (and still include your comment about named inputs)

File: allennlp/training/model.py (hunk 15 above)
Comment (DeNeutoy) [inline allennlp/training/model.py]: The API wrapper for the `nn.Sequential` stuff we talked about last week, where we wouldn't be able to use an allennlp model as a module unless we unpack the outputs.

File: allennlp/training/model.py (hunk 14 above)
Comment (DeNeutoy) [inline allennlp/training/model.py]: This way is closer to the pytorch api so I think this is preferable.

Comment (DeNeutoy) [inline allennlp/training/model.py]: It also saves unpacking the dictionary inside your forward call and makes it explicit that the `labels` are optional, as they can have `None` default arguments.

Comment (matt-gardner) [inline allennlp/training/model.py]: Ok, sounds good.
